server/controllers/v1/user.ts

```diff
@@ -16,6 +17,109 @@ export default class UserController {
     this.emailService = emailService;
   }
 
+  /**
+   * Update the last login time of user to the current date.
+   */
+  private updateLastLogin = async (user: User): Promise<void> => {
+    user.lastLogin = new Date();
+    await user.save();
+  };
+
+  /**
+   * Sign in the user with the given credentials if the user exists.
+   * Generate auth token and update last login date.
+   * @returns status code and data payload
+   */
+  async signIn(
+    userName: string,
+    password: string
+  ): Promise<{ status: number; data: { result?: User; message?: string } }> {
+    try {
+      let user = await this.userRepo.scope("withPassword").findOne({
+        where: {
+          userName: userName,
+        },
+      });
+      if (!user) {
+        return { status: 404, data: { message: "User doesn't exist" } };
+      }
+
+      const isPasswordCorrect = await argon2.verify(
+        user.password,
+        password,
+        {
+          type: argon2.argon2id,
+        }
+      );
+      if (!isPasswordCorrect) {
+        return { status: 400, data: { message: "Invalid credentials" } };
+      }
+
+      if (!user.confirmed) {
+        return { status: 403, data: { message: "Email has not been confirmed" }}
+      }
+
+      this.updateLastLogin(user);
+
+      user = await this.userRepo.findOne({
+        where: {
+          userName: userName,
+        },
+      });
```

Comment:
I don't think it's necessary to do a double query here. I would copy the original user object from the first query, delete the password field, and return that.